Serialize/ASN1.cs

```diff
@@ -21,6 +25,20 @@ public class ASN1
         /// </summary>
         public static Func<BacnetAddress, BacnetPropertyIds, byte, BacnetApplicationTags> CustomTagResolver;
 
+        private static readonly byte[] AnyDateOrTime = {0xFF, 0xFF, 0xFF, 0xFF};
+
+        /*
+         * TODO: I'm sorry, but I am going to paint a big, fat, "kill me" on you, IEncode / IDecode.
+         *
+         * There is this little thing known as "separation of concerns", and I don't think a class that
+         * holds data should be concernd with how (or even if) it gets serialized - that would make it
+         * very hard to have different encoders/decoders. We won't have different encoders because BACnet
+         * is a standard and there is only one way to encode/decode? Well, yes, except for application
+         * specifics and / or buggy hardware. Idealy, I'd be able to use a specific en/decoder per device
+         * I'm talking with, which inherits from a standard bacnet-decoder and can be extended / overridden.
+         *
+         * Also: cyclic dependencies. Bad thing.
+         */
```

Comment:
Ok, first of all I don't think we need a manifesto like this in the codebase. I think it's better to open up an issue and discuss it there. Just marking something for refactor or redesign like you did in other places would be better.

But to anwser it, of course I know what SoC is and I agree with your arguments (I don't know if this was directed to me or not but I anwser just in case). The problem is the legacy code I dind't write and that I didn't change much except of introducing those interfaces I think. I don't remember what I needed them for but for sure they made it easier dealing with this implementation.

Reply:
I'm sorry if that sounded like criticism of the work that was done there - that wasn't my intention.
I came across this code during debugging, and it was completely unexpected to find it implemented this way, because it was implemented differently everywhere else.

Since I didn't know if this was very old code that hasn't been refactored, a contribution from someone who just did things differently, or if it is intended to be the new way of doing things, I just wanted to mark it up for discussion.

The reasoning is there because I tend to forget why I marked something as TODO, and/or things look less bleak on a different day, so it's basically a note to myself :wink:

As for opening an issue: while splitting up the code and fixing issues, my impression was that there is much to do: inconsistent naming, duplicate code, strange ways of doing things. Some of those things changed back and forth a few times (sometimes I discovered why something was done this specific way), but I still think we may be better off re-imagining the whole encode/decode-stuff. I'm currently working on an experimental implementation to sharpen my idea of how this could be done differently.

Reply:
Sure, no harm done ;) At the beginning I hat little knowledge about BACnet protocol, but needed to communicate with BACnet device in order to integrate it. I liket the application YAME because it worked and was written in C# so decided to use its guts to do the job. I was trying not to break stuff at first but then the deeper I got into the code, the more frustrated I became, just like you I guess. So that you have a little perspective, please checkout and analyze the original code:

https://svn.code.sf.net/p/yetanotherbacnetexplorer/code/trunk/Yabe

Everything is stored mainly in BACnetBas.cs (8523 LoC) and BACnetClient.cs (2724 LoC). And I checked, it contains ASN1.IASN1encode interface but no ASN1.IASN1decode so I guess I added it and refactored the names. Anyway, I think you will like our codebase a little bit more after you analyze the original and think that I'm open to changing it ;)